### PR TITLE
Implement `DomainProvider` graph component

### PR DESCRIPTION
**Pull request by ancalita**
**Proposed changes**:
- implements #9332 

**Status (please check what you already did)**:
- [X] added some tests for the functionality
- [ ] updated the documentation
- [ ] updated the changelog (please check [changelog](https://github.com/RasaHQ/rasa/tree/main/changelog) for instructions)
- [X] reformat files using `black` (please check [Readme](https://github.com/RasaHQ/rasa#code-style) for instructions)

### CODE REVIEW

File: rasa/graph_components/providers/domain_provider.py [hunk 1]
```diff
@@ -0,0 +1,44 @@
...
+    def get_default_config() -> Dict[Text, Any]:
+        """Returns default configuration (see parent class for full docstring)."""
+        return {
+            "remove_duplicates": True,
+            "unique_last_num_states": None,
+            "augmentation_factor": 50,
+            "tracker_limit": None,
+            "use_story_concatenation": True,
+            "debug_plots": False,
+        }
```
Comment (twerkmeister) [inline rasa/graph_components/providers/domain_provider.py:28]: Where do these config parameters come from? They don't seem to be related to domains

File: rasa/graph_components/providers/domain_provider.py [hunk 2]
```diff
@@ -0,0 +1,44 @@
...
+from typing import Dict, Text, Any
+
+from rasa.engine.graph import GraphComponent, ExecutionContext
+from rasa.engine.storage.resource import Resource
+from rasa.engine.storage.storage import ModelStorage
+from rasa.shared.core.domain import Domain
+from rasa.shared.importers.importer import TrainingDataImporter
+
+
+class DomainProvider(GraphComponent):
```
Comment (twerkmeister) [inline rasa/graph_components/providers/domain_provider.py:11]: Reading the documentation I saw that the DomainProvider is marked as a persisting component. However, I also don't fully understand what should be persisted here and why ~ @wochinge is that persisting still relevant or did we remove that anyway by making the TrainingDataImporter fingerprint random?

**Review by twerkmeister — COMMENTED** [review]
Found two things :)

File: rasa/graph_components/providers/domain_provider.py (hunk 1 above)
Comment (ancalita) [reply rasa/graph_components/providers/domain_provider.py:28]: 😄 should have added a comment that because I wasn't sure if this component should have a config, I left these ones in from the template of `TrainingTrackerProvider` in hope to find out more during the review. I can remove this method if it's preferred to have an empty config.

Comment (wochinge) [reply rasa/graph_components/providers/domain_provider.py:28]: You can remove it as this one doesn't require a config 👍🏻

File: rasa/graph_components/providers/domain_provider.py [hunk 3]
```diff
@@ -0,0 +1,44 @@
...
+        config: Dict[Text, Any],
+        model_storage: ModelStorage,
+        resource: Resource,
+        execution_context: ExecutionContext,
+    ) -> DomainProvider:
+        """Creates component (see parent class for full docstring)."""
+        return cls(config)
+
+    @staticmethod
+    def generate_domain(importer: TrainingDataImporter) -> Domain:
```
Comment (wochinge) [inline rasa/graph_components/providers/domain_provider.py:42]: Could we call this `provide`? Makes it a bit simpler to generate the graph in the recipe if similar things have similar methods
```suggestion
    def provide(importer: TrainingDataImporter) -> Domain:
```

File: rasa/graph_components/providers/domain_provider.py (hunk 2 above)
Comment (wochinge) [reply rasa/graph_components/providers/domain_provider.py:11]: @twerkmeister Very good point! This component needs to be able to persist as this component also provides the domain during inference (our policies need the domain during inference). So the code should roughtly look like this:

```python

class DomainProvider:

   def __init__(domain: Optional):
    ...
    
   def provide_train(self, importer):
      domain = importer.get_domain()
      self.persist(self._resource, self._model_storage, domain)
      return domain
      
   def load(...):
     with model_storage(...):
        domain = read_domain_from_file)
     return cls(domain)
     
   def provide_inference(self):
      return self.domain
```

**Review by wochinge — CHANGES_REQUESTED** [review]

File: rasa/graph_components/providers/domain_provider.py (hunk 2 above)
Comment (ancalita) [reply rasa/graph_components/providers/domain_provider.py:11]: @wochinge which approach should I take for `persist` method 🤔 :
```
with self._model_storage.write(...) as directory:
        self.domain.persist(directory)
```
OR

```
domain_yaml = self.domain.as_yaml()
self._resource.to_cache(Path(domain_yaml), self._model_storage)
```

Comment (wochinge) [reply rasa/graph_components/providers/domain_provider.py:11]: the first one please 🙌🏻 

The second one should ideally never be called manually. That's done by the the `LocalTrainingCache` in the `TrainingHook`

File: rasa/graph_components/providers/domain_provider.py [hunk 4]
```diff
@@ -0,0 +1,51 @@
...
+        cls,
+        config: Dict[Text, Any],
+        model_storage: ModelStorage,
+        resource: Resource,
+        execution_context: ExecutionContext,
+    ) -> DomainProvider:
+        """Creates component (see parent class for full docstring)."""
+        with model_storage.read_from(resource) as resource_directory:
+            domain = Domain.from_path(resource_directory)
+        return cls(model_storage, resource, domain)
```
Comment (wochinge) [inline rasa/graph_components/providers/domain_provider.py:36]: this should be done in `load`. `create` is called during the training and in that case the `model_storage` won't contain an entry for this resource (it should even raise in this case 🤔 )

File: rasa/graph_components/providers/domain_provider.py [hunk 5]
```diff
@@ -0,0 +1,51 @@
...
+
+    def persist(self, domain: Domain) -> None:
+        """Persists domain to model storage."""
+        with self._model_storage.write_to(self._resource) as resource_directory:
+            domain.persist(resource_directory.joinpath("domain.yml"))
```
Comment (wochinge) [inline rasa/graph_components/providers/domain_provider.py:41]: optional: slightly fancier and (maybe) easier to read
```suggestion
            domain.persist(resource_director / "domain.yml")
```

File: rasa/graph_components/providers/domain_provider.py [hunk 6]
```diff
@@ -0,0 +1,51 @@
...
+    def __init__(
+        self,
+        model_storage: ModelStorage,
+        resource: Resource,
+        domain: Optional[Domain] = None,
+    ) -> None:
+        """Creates domain provider."""
+        self._model_storage = model_storage
+        self._resource = resource
+        self.domain = domain
```
Comment (wochinge) [inline rasa/graph_components/providers/domain_provider.py:23]:
```suggestion
        self._domain = domain
```

File: tests/test_graph_components/providers/test_domain_provider.py [hunk 7]
```diff
@@ -0,0 +1,37 @@
...
+
+
+def test_domain_provider_generates_and_persists_domain(
+    default_model_storage: ModelStorage,
+    default_execution_context: ExecutionContext,
+    tmp_path: Path,
+    config_path: Text,
+    domain_path: Text,
+):
+    resource = Resource.from_cache("xy", tmp_path, default_model_storage)
```
Comment (wochinge) [inline tests/test_graph_components/providers/test_domain_provider.py:19]: You can just use the constructor. The `from_cache` parts is basically always off limits as this is only called by the `TrainingHook`
```suggestion
    resource = Resource("some resource")
```

File: tests/test_graph_components/providers/test_domain_provider.py [hunk 8]
```diff
@@ -0,0 +1,37 @@
...
+        default_model_storage,
+        resource,
+        default_execution_context,
+    )
+    assert isinstance(component, DomainProvider)
+
+    importer = TrainingDataImporter.load_from_config(config_path, domain_path)
+    domain = component.provide_train(importer)
+
+    assert isinstance(domain, Domain)
```
Comment (wochinge) [inline tests/test_graph_components/providers/test_domain_provider.py:31]: can we also test `provide_inference`?

File: rasa/graph_components/providers/domain_provider.py [hunk 9]
```diff
@@ -0,0 +1,63 @@
...
+            domain = Domain.from_path(resource_directory)
+        return cls(model_storage, resource, domain)
+
+    def persist(self, domain: Domain) -> None:
+        """Persists domain to model storage."""
+        with self._model_storage.write_to(self._resource) as resource_directory:
+            domain.persist(resource_directory / "domain.yml")
+
+    def provide_train(self, importer: TrainingDataImporter) -> Domain:
+        """Generates loaded Domain of the bot during training."""
```
Comment (wochinge) [inline rasa/graph_components/providers/domain_provider.py:56]: As it doesn't "generate" anything I'd go with something like this? What do you think? 
```suggestion
        """Provides domain from training data during training."""
```

File: tests/test_graph_components/providers/test_domain_provider.py [hunk 10]
```diff
@@ -0,0 +1,43 @@
...
+
+    with default_model_storage.read_from(resource) as d:
+        match = list(d.glob("**/domain.yml"))
+        assert len(match) == 1
+        assert match[0].is_file()
+        assert domain.fingerprint() == Domain.from_path(match[0]).fingerprint()
```
Comment (wochinge) [inline tests/test_graph_components/providers/test_domain_provider.py:35]: can we maybe compare this to the content of the `domain` fixture? E.g. this wouldn't fail if we somehow persist the wrong domain to the `model_storage` and return the wrong domain in `provide_train` at the same time

File: rasa/graph_components/providers/domain_provider.py [hunk 11]
```diff
@@ -0,0 +1,63 @@
...
+        resource: Resource,
+        execution_context: ExecutionContext,
+        **kwargs: Any,
+    ) -> DomainProvider:
+        """Creates provider using a persisted version of itself."""
+        with model_storage.read_from(resource) as resource_directory:
+            domain = Domain.from_path(resource_directory)
+        return cls(model_storage, resource, domain)
+
+    def persist(self, domain: Domain) -> None:
```
Comment (wochinge) [inline rasa/graph_components/providers/domain_provider.py:50]: as it's not used outside of the class
```suggestion
    def _persist(self, domain: Domain) -> None:
```

**Review by wochinge — APPROVED** [review]
🎸

File: rasa/graph_components/providers/domain_provider.py (hunk 11 above)
Comment (ancalita) [reply rasa/graph_components/providers/domain_provider.py:50]: I also wanted to ask should I separate the tests to improve the coverage - atm that check fails because of coverage decrease?

Comment (wochinge) [reply rasa/graph_components/providers/domain_provider.py]: this test seems to be wrong - I just checked and you have a 100% test coverage. I fail to see how you could test better 😆 I'd just ignore the check for as it's not a required check.
